woocommerce/woocommerce
# Ensuring only single product creation with unique SKU for concurrent requests

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:
Closes #27295 



Added a method in the V2 product controller that check unique SKUs in the post meta before going ahead and creating the product.

### Before (on `trunk`)
![image](https://github.com/woocommerce/woocommerce/assets/15019298/27f3f48e-4b83-4b38-a27d-0dfafdf80f5f)

### After
![image](https://github.com/woocommerce/woocommerce/assets/15019298/b7128534-215c-4799-b454-26c83022fa77)




### How to test the changes in this Pull Request:


1. Copy the following script locally.
2. Now enter your credentials, you can get them from `woocommerce>settings>advanced>REST API`
3. Create new API keys if you don't have one already.
4. Add the CK and CS in the respective placeholders.
5. Replace the URL to the endpoint
6. Change the initial product SKU to something which is already not present on the table
7. Now run the script in `trunk` first, check the ` log_error_counts.txt` file, note the output
8. Check the products page in WC admin, you'll see multiple products created with the same SKU.
9. Run the script in this branch, check the ` log_error_counts.txt` file the output should be 1, 1
10. Check the products page in WC admin, you'll see only one product created with the same SKU.

I've created this bash script to test the change

```
#!/bin/bash

# Define the URL and authentication
URL="https://d.w.test/wp-json/wc/v3/products/batch"
CK="ck_7b87bfb00d141081ea0b2be4c443b939ac60677e"
CS="cs_0fbab889a09db5d83af5f2096ec553cf011c47c0"
CONTENT_TYPE="Content-Type: application/json"


INITIAL_SKU=4048

ROUNDS=5

CONCURRENT_REQUESTS=10

echo "Sending $ROUNDS rounds of $CONCURRENT_REQUESTS concurrent requests..."

count_errors() {
    local round=$1
    local error_count=0

    for i in $(seq 1 $CONCURRENT_REQUESTS); do
        if grep -q "error" "log_${round}_${i}.txt"; then
            error_count=$((error_count + 1))
        fi
    done

    echo $((CONCURRENT_REQUESTS - $error_count))
}

for j in $(seq 1 $ROUNDS); do
    SKU=$(($INITIAL_SKU + $j - 1)) 
    echo "Round $j with SKU $SKU..."

    for i in $(seq 1 $CONCURRENT_REQUESTS); do
        JSON_DATA=$(cat <<EOF
{
    "create": [
        {
            "name": "${j}-${i} New Test Product",
            "type": "simple",
            "sku": "${SKU}"
        }
    ]
}
EOF
)
        curl -X POST $URL \
            -u "$CK:$CS" \
            -H "$CONTENT_TYPE" \
            -d "$JSON_DATA" -k > "log_${j}_${i}.txt" 2>&1 &
    done

    wait
    echo "Completed round $j."

    error_count=$(count_errors $j)
    echo "Total products created in round $j with sku $SKU: $error_count" >> log_error_counts.txt
done

echo "All requests sent. Check log files for output."

```






### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
